src/src/com/microsoft/aad/adal/Oauth2.java

```diff
@@ -598,16 +593,13 @@ private AuthenticationResult processTokenResponse(HttpWebResponse webResponse) {
         case HttpURLConnection.HTTP_UNAUTHORIZED:
             try {
                 result = parseJsonResponse(webResponse.getBody());
-            } catch (final JSONException jsonException) {
-                Logger.e(TAG, jsonException.getMessage(), "", ADALError.SERVER_INVALID_JSON_RESPONSE, jsonException);
-                result = new AuthenticationResult(JSON_PARSING_ERROR, jsonException.getMessage(), null);
+            } catch (JSONException e) {
+                throw new AuthenticationException(ADALError.SERVER_INVALID_JSON_RESPONSE, "Can't parse server response " + webResponse.getBody(), e);
```

Comment:
I actually revert changes in this function, since it might end up in discussion and I'll submit it in separate PR